modules/gmake/src/proj.lua

```diff
@@ -0,0 +1,677 @@
+end
+
+
+---
+-- Prints the phonies of the project.
+--
+-- @param prj
+--  project to print phonies of
+---
+function proj.phonies(prj)
+	wl('.PHONY: clean prebuild')
+	wl()
+end
+
+
+---
+-- Prints the prebuild rules of the project.
+--
+-- @param prj
+--  project to print prebuild rules of
+---
+function proj.prebuildRule(prj)
+	wl('prebuild: | $(OBJDIR)')
+	export.indent()
+	wl('$(PREBUILDCMDS)')
+	export.outdent()
+	wl()
+	wl('$(OBJECTS): | prebuild')
+	wl()
+end
+
+
+---
+-- Prints the object directory rule.
+--
+-- @param prj
+--  project to print object directory rule of
+---
+function proj.objDirRule(prj)
+	wl('$(OBJDIR):')
+	export.indent()
+	wl('@echo "Creating $(OBJDIR)"')
+	export.outdent()
+	wl('ifeq (posix,$(SHELLTYPE))')
+	export.indent()
+	wl('$(SILENT) mkdir -p $(OBJDIR)')
+	export.outdent()
+	wl('else')
+	export.indent()
+	wl('$(SILENT) mkdir $(subst /,\\\\,$(OBJDIR))')
+	export.outdent()
+	wl('endif')
+	wl()
+end
+
+
+---
+-- Prints the target directory rule.
+--
+-- @param prj
+--  project to print target directory rule of
+---
+function proj.targetDirRule(prj)
+	wl('$(TARGETDIR):')
+	export.indent()
+	wl('@echo "Creating $(TARGETDIR)"')
+	export.outdent()
+	wl('ifeq (posix,$(SHELLTYPE))')
+	export.indent()
+	wl('$(SILENT) mkdir -p $(TARGETDIR)')
+	export.outdent()
+	wl('else')
+	export.indent()
+	wl('$(SILENT) mkdir $(subst /,\\\\,$(TARGETDIR))')
+	export.outdent()
+	wl('endif')
+	wl()
+end
+
+
+---
+-- Prints the target rule.
+--
+-- @param prj
+--  project to print target rule of
+---
+function proj.targetRule(prj)
+	wl('$(TARGET): $(OBJECTS) | $(TARGETDIR)')
+	export.indent()
+	wl('$(PRELINKCMDS)')
+	wl('@echo "Linking %s"', prj.name)
+	wl('$(SILENT) $(LINKCMD)')
+	wl('$(POSTBUILDCMDS)')
+	export.outdent()
+	wl()
+end
+
+
+---
+-- Prints the clean rule.
+--
+-- @param prj
+--  project to print clean rule of
+---
+function proj.cleanRule(prj)
+	wl('clean:')
+	export.indent()
+	wl('@echo "Cleaning %s"', prj.name)
+	export.outdent()
+	wl('ifeq (posix,$(SHELLTYPE))')
+	export.indent()
+	wl('$(SILENT) rm -f  $(TARGET)')
+	wl('$(SILENT) rm -rf $(OBJDIR)')
+	export.outdent()
+	wl('else')
+	export.indent()
+	wl('$(SILENT) if exist $(subst /,\\\\,$(TARGET)) del $(subst /,\\\\,$(TARGET))')
+	wl('$(SILENT) if exist $(subst /,\\\\,$(OBJDIR)) rmdir /s /q $(subst /,\\\\,$(OBJDIR))')
+	export.outdent()
+	wl('endif')
+	wl()
+end
+
+
+---
+-- Prints the all rule of the project.
+--
+-- @param prj
+--  project to print all rule of
+---
+function proj.allRule(prj)
+	wl('all: $(TARGET)')
+	export.indent()
+	wl('@:')
+	export.outdent()
+	wl()
+end
+
+
+---
+-- Prints the object file definition.
+--
+-- @param prj
+--  Project to print object files of
+---
+function proj.objects(prj)
+	wl('OBJECTS :=')
+	wl()
+
+	local files = {}
+
+	for _, file in ipairs(prj.files) do
+		if string.endsWith(file, proj.SOURCE_FILES) then
+			wl('OBJECTS += $(OBJDIR)/%s.o', path.getBaseName(file))
+		end
+	end
+	wl()
+end
```

Comment:
I think this could be split into project- and config-level lists, like you did for defines and the build flags.